components/places/src/storage/bookmarks/public_node.rs

```diff
@@ -111,27 +111,43 @@ fn get_child_guids(db: &PlacesDb, parent: RowId) -> Result<Vec<SyncGuid>> {
     )?)
 }
 
+enum ChildInfo {
+    NoChildren,
+    Guids(Vec<SyncGuid>),
+    Nodes(Vec<PublicNode>),
+}
+
+impl ChildInfo {
+    fn guids_nodes(self) -> (Option<Vec<SyncGuid>>, Option<Vec<PublicNode>>) {
```

Comment:
I was going to suggest having `ChildInfo` hold the parent, and implementing `From<ChildInfo> for PublicNode` that does what `PublicNode::from(rb).with_children(child_guids, child_nodes)` does now...but that's more invasive, so your call.

Reply:
This was the minimum to make it stop complaining about the types being too complex, we already have issues on file to clean this up more.